[VIVO-1755] - Better error handling when reasoner disabled

gneissone opened the pull request:
**[JIRA Issue](https://jira.lyrasis.org/browse/VIVO-1755)**: VIVO-1755

# What does this pull request do?
Fixes an issue that makes the site admin page inaccessible and the editor for data properties to break when the reasoner is disabled.

# What's new?
- Instead of throwing an error and stack trace up on the front end, the error is sent to the log. 
- An info box is displayed on the site admin page noting that the reasoner is disabled. The info box is already displayed when the reasoner status is abnormal, so I reused the functionality. 


# How should this be tested?
Reproduce the problem by commenting out these lines in tomcat/webapps/vivo/WEB-INF/resources/startup_listeners.txt

```
# edu.cornell.mannlib.vitro.webapp.application.ApplicationImpl$ReasonersSetup
# edu.cornell.mannlib.vitro.webapp.servlet.setup.SimpleReasonerSetup
```
Go to localhost:8080/vivo/siteAdmin.
You should get an error.

![Screen Shot 2020-03-20 at 12 45 51 PM](https://user-images.githubusercontent.com/10748475/77196350-cdeaf980-6aa8-11ea-80b6-5beef5b8038a.png)

Rebuild with changes. Site admin panel should work once again. Error should be printed to vivo.all.log, and stack trace should be printed to log if the log level is set to debug. 

![Screen Shot 2020-03-20 at 12 38 46 PM](https://user-images.githubusercontent.com/10748475/77196362-d5120780-6aa8-11ea-9381-f5c79033a81e.png)

Reproduce the data property editor error by turning reasoning off and trying to add a new record for a common data property, e.g. overview. An error will be thrown: java.lang.IllegalStateException: JFactTBoxReasonerModule has not been started. Rebuild with changes and you should be able to add the data property.

# Interested parties
@VIVO-project/vivo-committers

awoods commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/freemarker/BaseSiteAdminController.java:
```diff
@@ -199,12 +199,17 @@ protected ResponseValues processRequest(VitroRequest vreq) {
+                }
+            } catch (IllegalStateException e) {
+                error = "The inferencing engine is disabled.";
+                log.debug("Status of reasoner could not be determined.", e);
```
It seems like the debug message would be something more like the `error`:
```
Status of reasoner could not be determined because the inferencing engine is disabled.
```

awoods commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/VTwo/fields/IndividualsViaVClassOptions.java:
```diff
@@ -102,8 +107,13 @@ public FieldOptions setDefaultOptionLabel(String label){
+    	try {
+            TBoxReasonerStatus status = ApplicationUtils.instance().getTBoxReasonerModule().getStatus();
+            return status.isConsistent() && !status.isInErrorState();
+            } catch (IllegalStateException e) {
```
Minor: Formatting is a bit off.
```
} catch (...
```
..should be lined up with `try`.

gneissone commented:
> It appears that: `ApplicationUtils.instance().getTBoxReasonerModule()` is also being called at:
> https://github.com/vivo-project/Vitro/blob/master/api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/jena/JenaAdminActions.java#L195
> 
> ..it may be worth include a try/catch there as well.

Addressed your comments. I added the logic to JenaAdminActions.java but I would flag that servlet for possible removal, I've never seen it documented or used. I did test the various actions in the servlet and they appear to work, FWIW. I had to adjust the logging level from trace to info for the printRestrictions action to see the output since I wasn't about to set the entire application log to trace.

awoods commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/freemarker/BaseSiteAdminController.java:
```diff
@@ -199,12 +199,17 @@ protected ResponseValues processRequest(VitroRequest vreq) {
+                }
+            } catch (IllegalStateException e) {
+                error = "The inferencing engine is disabled.";
+                log.debug("Status of reasoner could not be determined. It is likely disabled.", e);
```
Suggest changing the logging level to `warn`.

gneissone replied:
I don't think this warrants a warning if the reasoner was intentionally disabled. I wouldn't want my log littered with stack traces if the application was running as I expected. If anything, perhaps a warning could be thrown during the startup tests if the reasoner isn't running if deemed necessary. Thoughts?

awoods replied:
Flagging in the startup-status would be great. Not changing the logging level here is fine.

gneissone replied:
Looked a little into this. It would be easy to copy the existing TBox reasoner check into a smoke test. Not sure if anything similar exists for the ABox (Simple) reasoner. The ABox reasoner doesn't appear to be abstracted out into an interface, so if I'm understanding things correctly we'd have to test for the Simple Reasoner explicitly, which could theoretically make it harder to swap out reasoners in the future. I believe this was done once in the past (Pellet -> Simple).

gneissone replied:
Per discussion on developer call today, will not add to startup smoke tests.

awoods commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/jena/JenaAdminActions.java:
```diff
@@ -192,29 +192,33 @@ private String testWriteXML() {
+			}
+		}
+		catch (IllegalStateException e) {
+			log.debug("Status of reasoner could not be determined. It is likely disabled", e);
```
Suggest changing the logging level to `warn`.

gneissone replied:
Ok, I agree this one deserves a warning since the function doesn't work at all if the reasoner is disabled.

awoods commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/dao/jena/DataPropertyDaoJena.java:
```diff
@@ -360,8 +360,13 @@ public List getAllExternalIdDataProperties() {
+            TBoxReasonerStatus status = ApplicationUtils.instance().getTBoxReasonerModule().getStatus();
+            return status.isConsistent() && !status.isInErrorState();
+        } catch (IllegalStateException e) {
+            log.debug("Status of reasoner could not be determined. It is likely disabled.", e);
```
Suggest changing the logging level to `warn`.

gneissone replied:
Same as above, I don't think this warrants a warn-level log because if the reasoner is unavailable it was likely intentional and the application will run as normal without it. This method is intended to determine if reasoning is available and returns false if not, that's not an error to me.

awoods commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/VTwo/fields/IndividualsViaVClassOptions.java:
```diff
@@ -102,8 +107,13 @@ public FieldOptions setDefaultOptionLabel(String label){
+            TBoxReasonerStatus status = ApplicationUtils.instance().getTBoxReasonerModule().getStatus();
+            return status.isConsistent() && !status.isInErrorState();
+        } catch (IllegalStateException e) {
+            log.debug("Status of reasoner could not be determined. It is likely disabled.", e);
```
Suggest changing the logging level to `warn`.

awoods commented:
Although I do not get the stacktrace once I deploy with this patch, I do not see the notification in the siteAdmin??
